common/persistence/cassandra/metadata_store.go

```diff
@@ -161,23 +168,48 @@ func (m *MetadataStore) CreateNamespaceInV2Table(
 
 	previous := make(map[string]interface{})
 	applied, iter, err := m.session.MapExecuteBatchCAS(batch, previous)
+	deleteOrphanNamespace := func() {
+		// Delete namespace from `namespaces_by_id`
+		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
+			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
+		}
+	}
+
 	if err != nil {
 		return nil, serviceerror.NewUnavailable(fmt.Sprintf("CreateNamespace operation failed. Inserting into namespaces table. Error: %v", err))
 	}
+
 	defer func() { _ = iter.Close() }()
 
 	if !applied {
-		// Namespace already exist.  Delete orphan namespace record before returning back to user
-		if errDelete := m.session.Query(templateDeleteNamespaceQuery, request.ID).WithContext(ctx).Exec(); errDelete != nil {
-			m.logger.Warn("Unable to delete orphan namespace record. Error", tag.Error(errDelete))
+
+		// if both conditions fail, find the one related to the first query
+		if name, ok := previous["name"]; !ok || name != request.Name {
+			m := make(map[string]interface{})
+			if iter.MapScan(m) {
+				previous = m
+			}
 		}
 
-		if id, ok := previous["Id"].([]byte); ok {
```

Comment:
It looks like the old logic was using "Id", not "id". Can you confirm which one is correct?

Reply:
the column name is `id`

Reply:
Oh. Can we fix that bug in a separate PR?

Reply:
Yeah it was a bug in the code and this code path would have never been reached. The type is also wrong. 

Is there a reason for fixing this in another PR? This condition return the same error as what comes after it with a different message.

Reply:
We need to separate the two fixes for better fault isolation. I'm not familiar with the code enough, but is it possible that a bug fix here could fix the larger issue this PR addresses?

Reply:
I can do that, but this does not fix the bug. The main bug we are fixing is in the other caller (`CreateNameSpace`) and what we are adding here is the side effect of letting namespaces create requests with the existing same Name and ID to be passed  This codepath only matters when we fix that bug. Otherwise, this falls back to return the same non-retryable error a few lines down with a different message. I am not sure how much difference does the message makes in handling this error, but I believe this should make a very big difference. WDYT?